### PR TITLE
Initial port of Chorus FX

### PR DESCRIPTION
**Pull request by jpjamipark**
Running into some bugs with the encoder values disappearing immediately + not updating properly until notes are pressed. Unsure if this is b/c of the port. Just putting this initial PR up for feedback since I'm new.

### CODE REVIEW

**Comment by jmidt** [comment]
I will take a closer look later, but I think you need to reset the envelopes in the actionhandlers in the screen. With a env.resetSoft()

File: src/engines/fx/chorus/audio.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  Audio::Audio() noexcept
+  {
+    // Set proper size of phase accumulator for graphics
+    phase.radius(1);
+  }
+
+  void Audio::action(itc::prop_change<&Props::delay>, float d) noexcept
+  {
+    //Goes from minimal to around 20 ms nominal delay.
```
Comment (topisani) [inline src/engines/fx/chorus/audio.cpp:15]: Please run `clang-format` on all files

File: src/engines/fx/chorus/audio.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include "core/voices/voice_manager.hpp"
+#include "util/dsp/chorus.hpp"
+
+namespace otto::engines::chorus {
+
+  using namespace core;
+
+  struct Audio {
+    Audio() noexcept;
+    float depth_ = 0.f;
```
Comment (topisani) [inline src/engines/fx/chorus/audio.hpp:16]: should be private, and down by the other variables

File: src/engines/fx/chorus/chorus.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  using namespace core::props;
+
+  struct Screen;
+  struct Audio;
+
+  using Sender = core::engine::EngineSender<Audio, Screen>;
+
+  struct Actions {
+    /// Publish the rotation variable, which is shared between the audio and screen
```
Comment (topisani) [inline src/engines/fx/chorus/chorus.hpp:17]: copy paste error in doc comment: `rotation`

File: src/engines/fx/chorus/chorus.hpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  struct Chorus : core::engine::EffectEngine<Chorus> {
+    static constexpr util::string_ref name = "Chorus";
+    Chorus();
+
+    void encoder(core::input::EncoderEvent e) override;
+
+    core::ui::ScreenAndInput screen() override;
+  
+    std::unique_ptr<Audio> audio;
+    std::unique_ptr<Screen> screen_;
```
Comment (topisani) [inline src/engines/fx/chorus/chorus.hpp:40]: everything below audio can be private

File: src/engines/fx/chorus/screen.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    float spacing = spacing_constant * delay_ + 10;
+    Point start = {120 - delay_ * 50, 165};
+
+    OTTO_ASSERT(phase_ != nullptr);
+
+    for (int i=num_heads; i>=1; i--) {
+      float head_height = wave_height * gam::scl::sinP9(gam::scl::wrap(*phase_ - 0.2f*(float)i, 1.f, -1.f));
+      //This brightness calculation is HORRIBLY over-engineered...
+      float brightness = powf(cosh(abs(feedback_)*1.5)*0.4, 0.3 * float(i));
```
Comment (topisani) [inline src/engines/fx/chorus/screen.cpp:107]: Prefer to move variables like `brightness` (and anything else that is computed solely based on the properties) out to member variables, and compute them in the action handlers.

File: src/engines/fx/chorus/screen.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    float depth_ = 0.f;
+    void action(itc::prop_change<&Props::depth>, float d) noexcept;
+    float feedback_ = 0.f;
+    void action(itc::prop_change<&Props::feedback>, float f) noexcept;
+    float rate_ = 0.f;
+    void action(itc::prop_change<&Props::rate>, float r) noexcept;
+    std::atomic<float>* phase_ = nullptr;
+    void action(Actions::phase_value, std::atomic<float>& v) noexcept;
+
+    std::array<ui::vg::Color, 10> colour_list = {ui::vg::Colors::Blue,
```
Comment (topisani) [inline src/engines/fx/chorus/screen.hpp:30]: this array should be constexpr and declared in the .cpp file. `vg::Color` is `constexpr`-enabled on `develop` now, so this should be doable if you pull from that.

File: src/engines/fx/chorus/screen.hpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+namespace otto::engines::chorus {
+
+  using namespace core;
+
+  struct Screen : ui::Screen {
+    void draw(nvg::Canvas& ctx) override;
+    void draw_front_head(nvg::Canvas&, nvg::Point, ui::vg::Color, float);
+    void draw_background_head(nvg::Canvas&, nvg::Point, ui::vg::Color, float);
+
+    float delay_ = 0.f;
```
Comment (topisani) [inline src/engines/fx/chorus/screen.hpp:19]: all of these variables should be private and declared down by the other variables.

**Review by topisani — CHANGES_REQUESTED** [review]

File: src/engines/fx/chorus/chorus.hpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    
+
+    std::atomic<float> phase_ = 0;
+
+    Sender sender_ = {*audio, *screen_};
+    Props props;
+  };
+
+} // namespace otto::engines::chorus
```
Comment (topisani) [inline src/engines/fx/chorus/chorus.hpp:49]:
```suggestion
} // namespace otto::engines::chorus
#include "audio.hpp"
#include "screen.hpp"
```
To fix travis errors

File: src/engines/fx/chorus/screen.cpp (hunk 5 above)
Comment (jpjamipark) [reply src/engines/fx/chorus/screen.cpp:107]: It looks like `brightness` &  `head_height` also rely on the i value based on the head - so should I create an array of size `num_heads` and store brightness/head_height for each head ?

**Comment by jmidt** [comment]
Yes, that sounds fine for now!

man. 2. dec. 2019 02.35 skrev jpjamipark <notifications@github.com>:

File: src/engines/fx/chorus/screen.cpp (hunk 5 above)
Comment (topisani) [reply src/engines/fx/chorus/screen.cpp:107]: yep, thats a great idea!

**Comment by jmidt** [comment]
Merging this. I will finish this now.

**Pull request merged**